hazelcast/src/test/java/com/hazelcast/client/map/ClientMapIterableTest.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.map;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.impl.proxy.ClientMapProxy;
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.spi.properties.ClusterProperty;
+import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ClientMapIterableTest extends HazelcastTestSupport {
```

Comment:
Since there are lots of test duplication, what do you think about something like the following structure?
```java
public abstract class AbstractMapIterableTest extends HazelcastTestSupport {

    protected TestHazelcastFactory factory = new TestHazelcastFactory();
    protected HazelcastInstance proxyInstance;
    private HazelcastInstance server;

    @Parameterized.Parameter
    public boolean prefetchValues;

    @Parameterized.Parameters(name = "prefetchValues:{0}")
    public static Iterable<Object[]> parameters() {
        return asList(new Object[] {Boolean.TRUE}, new Object[] {Boolean.FALSE});
    }

    @Before
    public void setup() {
        server = factory.newHazelcastInstance(smallInstanceConfig());
        proxyInstance = getProxyInstance();
    }

    @After
    public void teardown() {
        factory.terminateAll();
    }

    protected abstract <K, V> Iterable getIterable(IMap<K, V> map);
    protected abstract HazelcastInstance getProxyInstance();

    @Test(expected = IllegalStateException.class)
    public void test_remove_Throws_Exception_When_Called_Without_Next() {
        IMap<Object, Object> map = proxyInstance.getMap(randomName());
        Iterator<Map.Entry<Object, Object>> iterator = getIterable(map).iterator();
        iterator.remove();
    }
}
```

```java
@RunWith(Parameterized.class)
@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
@Category({QuickTest.class, ParallelJVMTest.class})
public class ClientMapIterableTest extends AbstractMapIterableTest {
    @Override
    protected <K, V> Iterable getIterable(IMap<K, V> map) {
        return ((ClientMapProxy<K, V>) map).iterable(10, prefetchValues);
    }

    @Override
    protected HazelcastInstance getProxyInstance() {
        return factory.newHazelcastClient();
    }
}
```

and the similar one with the server side proxy